File: go-le/bplus_tree.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

const MinDegree = 3

// Entry represents a key-value pair
type Entry struct {
	Key   int
	Value string
}

// BPlusNode represents a node in the B+ tree
type BPlusNode struct {
	IsLeaf   bool
	Entries  []Entry      // For leaf nodes
	Keys     []int        // For internal nodes
	Children []*BPlusNode // For internal nodes
}

// BPlusTree represents the B+ tree
type BPlusTree struct {
	Root   *BPlusNode
	Height int
}

// NewBPlusTree creates a new empty B+ tree
func NewBPlusTree() *BPlusTree {
	return &BPlusTree{
		Root: &BPlusNode{
			IsLeaf:  true,
			Entries: make([]Entry, 0),
		},
		Height: 1,
	}
}

// NewLeafNode creates a new leaf node
func NewLeafNode() *BPlusNode {
	return &BPlusNode{
		IsLeaf:  true,
		Entries: make([]Entry, 0),
	}
}

// NewInternalNode creates a new internal node
func NewInternalNode() *BPlusNode {
	return &BPlusNode{
		IsLeaf:   false,
		Keys:     make([]int, 0),
		Children: make([]*BPlusNode, 0),
	}
}

// IsFull checks if a node is full
func (n *BPlusNode) IsFull() bool {
	if n.IsLeaf {
		return len(n.Entries) >= 2*MinDegree-1
	}
	return len(n.Keys) >= 2*MinDegree-1
}

// NumKeys returns the number of keys in the node
func (n *BPlusNode) NumKeys() int {
	if n.IsLeaf {
		return len(n.Entries)
	}
	return len(n.Keys)
}

// Insert inserts a key-value pair into the tree
func (t *BPlusTree) Insert(key int, value string) {
	if t.Root.IsFull() {
		oldRoot := t.Root
		t.Root = NewInternalNode()
		t.Root.Children = append(t.Root.Children, oldRoot)
		t.splitChild(0)
		t.Height++
	}

	t.insertNonFull(key, value)
}

// insertNonFull inserts into a node that is not full
func (t *BPlusTree) insertNonFull(key int, value string) {
	node := t.Root

	for !node.IsLeaf {
		// Find the child where the key should go
		childIdx := 0
		for i, k := range node.Keys {
			if key < k {
				childIdx = i
				break
			}
			childIdx = i + 1
		}

		// Check if child is full
		if node.Children[childIdx].IsFull() {
			t.splitChildInternal(node, childIdx)
			if key > node.Keys[childIdx] {
				childIdx++
			}
		}

		node = node.Children[childIdx]
	}

	// Insert into leaf
	idx := sort.Search(len(node.Entries), func(i int) bool {
		return node.Entries[i].Key > key
	})

	// Check if key exists
	if idx > 0 && node.Entries[idx-1].Key == key {
		node.Entries[idx-1].Value = value
	} else {
		// Insert new entry
		newEntry := Entry{Key: key, Value: value}
		node.Entries = append(node.Entries[:idx], append([]Entry{newEntry}, node.Entries[idx:]...)...)
	}
}

// splitChild splits the child of the root
func (t *BPlusTree) splitChild(childIdx int) {
	oldChild := t.Root.Children[0]
	mid := MinDegree - 1

	if oldChild.IsLeaf {
		// Split leaf node
		newLeaf := NewLeafNode()

		// Copy second half to new leaf
		newLeaf.Entries = make([]Entry, len(oldChild.Entries)-mid)
		copy(newLeaf.Entries, oldChild.Entries[mid:])

		// Keep first half in old leaf
		oldChild.Entries = oldChild.Entries[:mid]

		// Add middle key to root
		t.Root.Keys = append(t.Root.Keys, newLeaf.Entries[0].Key)
		t.Root.Children = append(t.Root.Children, newLeaf)
	}
}

// splitChildInternal splits a child of an internal node
func (t *BPlusTree) splitChildInternal(parent *BPlusNode, childIdx int) {
	child := parent.Children[childIdx]
	mid := MinDegree - 1

	if child.IsLeaf && len(child.Entries) > mid {
		newLeaf := NewLeafNode()

		// Copy second half to new leaf
		newLeaf.Entries = append(newLeaf.Entries, child.Entries[mid:]...)

		// Keep first half in old child
		child.Entries = child.Entries[:mid]

		// Insert middle key into parent
		middleKey := newLeaf.Entries[0].Key

		// Insert key and child into parent
		parent.Keys = append(parent.Keys, 0)
		copy(parent.Keys[childIdx+1:], parent.Keys[childIdx:])
		parent.Keys[childIdx] = middleKey

		parent.Children = append(parent.Children, nil)
		copy(parent.Children[childIdx+2:], parent.Children[childIdx+1:])
		parent.Children[childIdx+1] = newLeaf
	}
}

// Search searches for a value by key
func (t *BPlusTree) Search(key int) (string, bool) {
	return t.searchRecursive(t.Root, key)
}

// searchRecursive recursively searches for a key
func (t *BPlusTree) searchRecursive(node *BPlusNode, key int) (string, bool) {
	if node.IsLeaf {
		// Linear search in leaf
		for _, entry := range node.Entries {
			if entry.Key == key {
				return entry.Value, true
			}
		}
		return "", false
	}

	// Find the child where the key might be
	childIdx := 0
	for i, k := range node.Keys {
		if key < k {
			childIdx = i
			break
		}
		childIdx = i + 1
	}

	return t.searchRecursive(node.Children[childIdx], key)
}

// RangeQuery finds all entries in range [start, end]
func (t *BPlusTree) RangeQuery(start, end int) []Entry {
	var result []Entry
	t.rangeQueryRecursive(t.Root, start, end, &result)
	return result
}

// rangeQueryRecursive recursively searches for keys in range
func (t *BPlusTree) rangeQueryRecursive(node *BPlusNode, start, end int, result *[]Entry) {
	if node.IsLeaf {
		for _, entry := range node.Entries {
			if entry.Key >= start && entry.Key <= end {
				*result = append(*result, entry)
			}
		}
		return
	}

	// Search in children
	for i, key := range node.Keys {
		if start <= key {
			t.rangeQueryRecursive(node.Children[i], start, end, result)
		}
	}

	// Check last child
	if len(node.Children) > len(node.Keys) {
		lastKey := 0
		if len(node.Keys) > 0 {
			lastKey = node.Keys[len(node.Keys)-1]
		}
		if end > lastKey {
			t.rangeQueryRecursive(node.Children[len(node.Keys)], start, end, result)
		}
	}
}

// AllKeys returns all keys in sorted order
func (t *BPlusTree) AllKeys() []int {
	var keys []int
	t.collectKeys(t.Root, &keys)
	return keys
}

// collectKeys recursively collects all keys
func (t *BPlusTree) collectKeys(node *BPlusNode, keys *[]int) {
	if node.IsLeaf {
		for _, entry := range node.Entries {
			*keys = append(*keys, entry.Key)
		}
	} else {
		for _, child := range node.Children {
			t.collectKeys(child, keys)
		}
	}
}

// PrintTree prints the tree structure
func (t *BPlusTree) PrintTree() {
	fmt.Printf("B+ Tree (min_degree = %d)\n", MinDegree)
	fmt.Printf("Height: %d\n", t.Height)
	t.printNode(t.Root, 0)
}

// printNode recursively prints a node
func (t *BPlusTree) printNode(node *BPlusNode, level int) {
	indent := strings.Repeat("  ", level)

	if node.IsLeaf {
		keys := make([]int, len(node.Entries))
		for i, entry := range node.Entries {
			keys[i] = entry.Key
		}
		fmt.Printf("%sLeaf: %v\n", indent, keys)
		for _, entry := range node.Entries {
			fmt.Printf("%s  %d -> %s\n", indent, entry.Key, entry.Value)
		}
	} else {
		fmt.Printf("%sInternal: %v\n", indent, node.Keys)
		for _, child := range node.Children {
			t.printNode(child, level+1)
		}
	}
}

// String returns a string representation of the tree
func (t *BPlusTree) String() string {
	return fmt.Sprintf("B+ Tree(height=%d, keys=%v)", t.Height, t.AllKeys())
}

File: rusty-le/src/bplus_tree.rs
use std::fmt;

const MIN_DEGREE: usize = 3;

/// B+ Tree Entry with key and value
#[derive(Clone, Debug)]
pub struct Entry {
    pub key: i32,
    pub value: String,
}

/// B+ Tree Node - either Leaf or Internal
#[derive(Clone, Debug)]
pub enum Node {
    Leaf {
        entries: Vec<Entry>,
    },
    Internal {
        keys: Vec<i32>,
        children: Vec<Box<Node>>,
    },
}

impl Node {
    pub fn new_leaf() -> Self {
        Node::Leaf {
            entries: Vec::new(),
        }
    }

    pub fn new_internal() -> Self {
        Node::Internal {
            keys: Vec::new(),
            children: Vec::new(),
        }
    }

    pub fn is_leaf(&self) -> bool {
        matches!(self, Node::Leaf { .. })
    }

    pub fn num_keys(&self) -> usize {
        match self {
            Node::Leaf { entries } => entries.len(),
            Node::Internal { keys, .. } => keys.len(),
        }
    }

    pub fn is_full(&self) -> bool {
        self.num_keys() >= 2 * MIN_DEGREE - 1
    }
}

/// B+ Tree Implementation
pub struct BPlusTree {
    root: Box<Node>,
    height: usize,
}

impl BPlusTree {
    /// Create a new empty B+ Tree
    pub fn new() -> Self {
        BPlusTree {
            root: Box::new(Node::new_leaf()),
            height: 1,
        }
    }

    /// Insert a key-value pair
    pub fn insert(&mut self, key: i32, value: String) {
        if self.root.is_full() {
            let old_root = std::mem::replace(&mut self.root, Box::new(Node::new_internal()));
            
            if let Node::Internal {
                keys: _,
                ref mut children,
            } = *self.root
            {
                children.push(old_root);
                self.split_child(0);
            }
            
            self.height += 1;
        }

        self.insert_non_full(key, value);
    }

    fn insert_non_full(&mut self, key: i32, value: String) {
        if self.root.is_leaf() {
            if let Node::Leaf { ref mut entries } = *self.root {
                if let Some(pos) = entries.iter().position(|e| e.key == key) {
                    entries[pos].value = value;
                } else {
                    let pos = entries.iter().position(|e| e.key > key).unwrap_or(entries.len());
                    entries.insert(pos, Entry { key, value });
                }
            }
        } else {
            let mut child_idx = 0;
            {
                if let Node::Internal { keys, .. } = self.root.as_ref() {
                    for (i, k) in keys.iter().enumerate() {
                        if key < *k {
                            child_idx = i;
                            break;
                        }
                        child_idx = i + 1;
                    }
                }
            }

            let should_split = if let Node::Internal { children, .. } = self.root.as_ref() {
                children[child_idx].is_full()
            } else {
                false
            };

            if should_split {
                self.split_child_internal(child_idx);
                let key_val = if let Node::Internal { keys, .. } = self.root.as_ref() {
                    keys[child_idx]
                } else {
                    i32::MIN
                };
                if key > key_val {
                    child_idx += 1;
                }
            }

            self.insert_into_child(child_idx, key, value);
        }
    }

    fn insert_into_child(&mut self, child_idx: usize, key: i32, value: String) {
        if let Node::Internal { ref mut children, .. } = *self.root {
            if children[child_idx].is_leaf() {
                if let Node::Leaf { ref mut entries } = *children[child_idx] {
                    if let Some(pos) = entries.iter().position(|e| e.key == key) {
                        entries[pos].value = value;
                    } else {
                        let pos = entries.iter().position(|e| e.key > key).unwrap_or(entries.len());
                        entries.insert(pos, Entry { key, value });
                    }
                }
            }
        }
    }

    fn split_child(&mut self, child_idx: usize) {
        if let Node::Internal {
            ref mut keys,
            ref mut children,
        } = *self.root
        {
            let mid = MIN_DEGREE - 1;
            let child = children[child_idx].clone();

            if let Node::Leaf { ref entries } = child.as_ref() {
                if entries.len() > mid {
                    let split_key = entries[mid].key;
                    
                    let left_entries = entries[..mid].to_vec();
                    let right_entries = entries[mid..].to_vec();

                    children[child_idx] = Box::new(Node::Leaf {
                        entries: left_entries,
                    });

                    let right_child = Box::new(Node::Leaf {
                        entries: right_entries,
                    });

                    keys.insert(child_idx, split_key);
                    children.insert(child_idx + 1, right_child);
                }
            }
        }
    }

    fn split_child_internal(&mut self, child_idx: usize) {
        if let Node::Internal {
            ref mut keys,
            ref mut children,
        } = *self.root
        {
            let mid = MIN_DEGREE - 1;
            let child = children[child_idx].clone();

            if let Node::Leaf { ref entries } = child.as_ref() {
                if entries.len() > mid {
                    let split_key = entries[mid].key;
                    
                    let left_entries = entries[..mid].to_vec();
                    let right_entries = entries[mid..].to_vec();

                    children[child_idx] = Box::new(Node::Leaf {
                        entries: left_entries,
                    });

                    let right_child = Box::new(Node::Leaf {
                        entries: right_entries,
                    });

                    keys.insert(child_idx, split_key);
                    children.insert(child_idx + 1, right_child);
                }
            }
        }
    }

    /// Search for a value by key
    pub fn search(&self, key: i32) -> Option<String> {
        self.search_recursive(&self.root, key)
    }

    fn search_recursive(&self, node: &Node, key: i32) -> Option<String> {
        match node {
            Node::Leaf { entries } => {
                entries
                    .iter()
                    .find(|e| e.key == key)
                    .map(|e| e.value.clone())
            }
            Node::Internal { keys, children } => {
                let mut child_idx = 0;
                for (i, k) in keys.iter().enumerate() {
                    if key < *k {
                        child_idx = i;
                        break;
                    }
                    child_idx = i + 1;
                }
                self.search_recursive(&children[child_idx], key)
            }
        }
    }

    /// Range query: find all entries in range [start, end]
    pub fn range_query(&self, start: i32, end: i32) -> Vec<(i32, String)> {
        let mut result = Vec::new();
        self.range_query_recursive(&self.root, start, end, &mut result);
        result
    }

    fn range_query_recursive(
        &self,
        node: &Node,
        start: i32,
        end: i32,
        result: &mut Vec<(i32, String)>,
    ) {
        match node {
            Node::Leaf { entries } => {
                for entry in entries {
                    if entry.key >= start && entry.key <= end {
                        result.push((entry.key, entry.value.clone()));
                    }
                }
            }
            Node::Internal { keys, children } => {
                for (i, key) in keys.iter().enumerate() {
                    if start <= *key {
                        self.range_query_recursive(&children[i], start, end, result);
                    }
                }
                if let Some(last_child) = children.last() {
                    if end > keys.last().copied().unwrap_or(i32::MIN) {
                        self.range_query_recursive(last_child, start, end, result);
                    }
                }
            }
        }
    }

    /// Get all keys in sorted order
    pub fn all_keys(&self) -> Vec<i32> {
        let mut keys = Vec::new();
        self.collect_keys(&self.root, &mut keys);
        keys
    }

    fn collect_keys(&self, node: &Node, keys: &mut Vec<i32>) {
        match node {
            Node::Leaf { entries } => {
                for entry in entries {
                    keys.push(entry.key);
                }
            }
            Node::Internal { children, .. } => {
                for child in children {
                    self.collect_keys(child, keys);
                }
            }
        }
    }

    /// Print tree structure
    pub fn print_tree(&self) {
        println!("B+ Tree (min_degree = {})", MIN_DEGREE);
        println!("Height: {}", self.height);
        self.print_node(&self.root, 0);
    }

    fn print_node(&self, node: &Node, level: usize) {
        let indent = "  ".repeat(level);
        match node {
            Node::Leaf { entries } => {
                let keys: Vec<i32> = entries.iter().map(|e| e.key).collect();
                println!("{}Leaf: {:?}", indent, keys);
                for entry in entries {
                    println!("{}  {} -> {}", indent, entry.key, entry.value);
                }
            }
            Node::Internal { keys, children } => {
                println!("{}Internal: {:?}", indent, keys);
                for child in children {
                    self.print_node(child, level + 1);
                }
            }
        }
    }
}

impl fmt::Display for BPlusTree {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "B+ Tree(height={}, keys={:?})",
            self.height,
            self.all_keys()
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_insert_and_search() {
        let mut tree = BPlusTree::new();
        tree.insert(10, "ten".to_string());
        tree.insert(20, "twenty".to_string());
        tree.insert(5, "five".to_string());

        assert_eq!(tree.search(10), Some("ten".to_string()));
        assert_eq!(tree.search(20), Some("twenty".to_string()));
        assert_eq!(tree.search(5), Some("five".to_string()));
        assert_eq!(tree.search(100), None);
    }

    #[test]
    fn test_range_query() {
        let mut tree = BPlusTree::new();
        for i in 1..=10 {
            tree.insert(i * 10, format!("value_{}", i));
        }

        let result = tree.range_query(25, 75);
        assert!(!result.is_empty());
    }
}
